schism/compiler.ss

```diff
@@ -953,6 +1015,98 @@
+  ;; a string, but eventually we will be able to do this ahead of time
+  ;; and store the strings in a static data section.
+
+  (define (lower-strings fn*)
+    (map (lambda (fn) (lower-strings-fn fn)) fn*))
+  (define (lower-strings-fn fn)
+    (if (eq? (car fn) '%wasm-import)
+        fn
+        (let ((def (car fn))
+              (body (cadr fn)))
+          `(,def ,(lower-strings-expr body)))))
+  (define (lower-strings-expr expr)
+    (if (and (pair? expr) (symbol? (car expr)))
+        (let ((tag (car expr)))
+          (cond
+           ((eq? tag 'null) expr)
+           ((eq? tag 'bool) expr)
+           ((eq? tag 'char) expr)
+           ((eq? tag 'begin)
+	    (cons 'begin (map lower-strings-expr (cdr expr))))
+           ((eq? tag 'var) expr)
+           ((eq? tag 'number) expr)
+           ((eq? tag 'eof-object) expr)
+	   ((eq? tag 'string) (lower-string (cadr expr)))
+           ((eq? tag 'call)
+            (cons 'call (cons (cadr expr)
+			      (map lower-strings-expr (cddr expr)))))
+           ((eq? tag 'apply-procedure)
+            (cons 'apply-procedure (map lower-strings-expr (cdr expr))))
+           ((eq? tag 'if)
+            (let ((t (cadr expr))
+                  (c (caddr expr))
+                  (a (cadddr expr)))
+              `(if
+                ,(lower-strings-expr t)
+                ,(lower-strings-expr c)
+                ,(lower-strings-expr a))))
+           ((eq? tag 'let)
+            `(let ,(lower-strings-bindings (cadr expr))
+               ,(lower-strings-expr (caddr expr))))
+           ((eq? tag '%function-index) expr)
+           ((eq? tag 'icall)
+            (cons tag (cons (cadr expr) (map lower-strings-expr (cddr expr)))))
+	   ((eq? tag 'zero?) `(zero? ,(lower-strings-expr (cadr expr))))
+	   ((eq? tag 'eq?) `(eq?
+			     ,(lower-strings-expr (cadr expr))
+			     ,(lower-strings-expr (caddr expr))))
+	   ((eq? tag '<) `(<
+			   ,(lower-strings-expr (cadr expr))
+			   ,(lower-strings-expr (caddr expr))))
+	   ((eq? tag 'neq?) `(neq?
+			      ,(lower-strings-expr (cadr expr))
+			      ,(lower-strings-expr (caddr expr))))
+	   ((eq? tag 'bool) expr)
+           (else (begin (write "unrecognized expr: ")
+                        (display expr)
+                        (newline)
+                        (error 'lower-strings-expr "Unrecognized expr")))))
+        (begin (write "error in lower-strings-expr: `")
+               (display expr)
+               (write "` is malformed")
+               (newline)
+               (error 'lower-strings-expr "malformed expr"))))
+  (define (lower-strings-bindings bindings)
+    (map (lambda (b)
+	   (let ((name (car b))
+		 (expr (cadr b)))
+	     `(,name ,(lower-strings-expr expr))))
+	 bindings))
+  (define (lower-string s)
+    (let ((chars (string->list s))
+	  (string-var (gensym "string")))
+      `(let ((,string-var (call make-string (number ,(length chars)))))
+	 (begin
+           (begin . ,(for-index (lambda (i c)
```

Comment:
Might as well avoid the list and make for-index be vector-for-each or something.